docs/architecture/adr/0022-server-authorization.md

```diff
@@ -0,0 +1,244 @@
+    before an authorization decision can be made. This makes it impractical for large read
+    operations, such as syncing a user's vault, as we cannot realistically fetch _all_ ciphers from
+    the database and use `AuthorizationService` filter them. Therefore, some high frequency/high
+    cost operations would continue to use authorization logic embedded in database queries
+  - the authorization handlers may need to fetch additional data from the database to make a
+    decision. This may cause many database reads, particularly when iterating over multiple items,
+    unless an additional caching solution is implemented. This is probably manageable for now by
+    caching frequently needed data in `CurrentContext` or the handler itself, both of which are
+    scoped to the lifetime of the request
+- for the reasons above, it is not a complete solution: it would cover _most_ but not all of our use
+  cases
+- server-side solution only - client code (particularly in Admin Console) needs to continue to
+  maintain its own duplicate logic to determine what UI flows it should show/hide and enable/disable
+- a server endpoint needs to understand what resources are affected, which can be nuanced when
+  dealing with relational data (e.g. saving a vault item may require separate authorization checks
+  for the `Cipher` resource and the `CollectionCipher` relationship). However, this may be an issue
+  for other solutions as well, as our endpoints often update multiple resources at once
+
+### [OpenFGA](https://openfga.dev)
+
+#### Summary
+
+OpenFGA is an open-source implementation of Google's Zanzibar authorization system, which is used
+for Google Docs. It uses
+[Relationship Based Access Control (ReBAC)](https://openfga.dev/docs/modeling/getting-started) which
+makes authorization decisions based on a user's relationship to an object (resource):
+
+> Authorization decisions are then yes or no answers to the question: "Does user U have relation R
+> with object O?".
+
+To start using OpenFGA:
+
+- you define your authorization scheme (in terms of users, objects, and relationships) in its DSL
+- you seed the OpenFGA store (database) with your current users, objects, and relationships, which
+  it will use to make authorization decisions
+- you keep the OpenFGA store up-to-date with any changes in your underlying data (e.g. as users are
+  added and removed). Changes to the Bitwarden database _do not_ automatically update the OpenFGA
+  store
+
+OpenFGA then supports the following queries:
+
+- "check requests" (does user U have relation R with object O?) - for simple authorization checks
+- "list objects requests" (get all objects with which user U has relationship R) - for bulk read
+  operations
+- "list users requests" (get all users that have relationship R with object O)
+
+OpenFGA has a [.NET SDK](https://github.com/openfga/dotnet-sdk) and a
+[Javascript SDK](https://github.com/openfga/js-sdk) with Typescript types.
+
+#### Advantages
```

Comment:
ℹ️ Interesting case study to me: https://medium.com/building-carta/authz-cartas-highly-scalable-permissions-system-782a7f2c840f